src/DonationForms/Listeners/StoreCustomFields.php

```diff
@@ -21,20 +25,56 @@ public function __invoke(DonationForm $form, Donation $donation, array $customFi
     {
         $form->schema()->walkFields(
             function (Field $field) use ($customFields, $donation) {
-                if (!array_key_exists($field->getName(), $customFields)) {
+                $fieldName = $field->getName();
+                $fieldType = $field->getType();
+
+                if (!array_key_exists($fieldName, $customFields)) {
                     return;
                 }
 
-                $value = $customFields[$field->getName()];
-
-                if ($field->shouldStoreAsDonorMeta()) {
-                    // save as donor meta
-                    give()->donor_meta->update_meta($donation->donorId, $field->getName(), $value);
+                /** @var File $field */
+                if (($fieldType === Types::FILE) && isset($_FILES[$fieldName])) {
+                    //TODO: let file field provide this storage logic
+                    $this->handleFileUpload($field, $donation);
                 } else {
-                    // save as donation meta
-                    give()->payment_meta->update_meta($donation->id, $field->getName(), $value);
+                    $value = $customFields[$fieldName];
+
+                    // TODO: it would be nice to have a way of serializing the value from the field api
+                    $value = apply_filters("givewp_store_custom_field_value_$fieldName", $value, $field, $donation);
```

Comment:
Serializing the value in what context?

Reply:
The context is for storage. Suppose you are sending an array to the server but want to store it as a pipe delimited string.  Right now the only way to do that would be to use a custom validation rule and call `serialize()`.  This might be a non issue with a fields api storage scope thing but could also be useful if you just want to opt into saving as donation meta but modify the value before storing.

Reply:
Validation rules are capable of sanitizing, is that sufficient? Or we're thinking more?

Reply:
Yeah that's how we're getting around those now which is sufficient for the time being.  Can circle back to this discussion